File: 4 Kyu/4 kyu - Longest Common Subsequence (Performance version).cpp
#include <string>
#include <algorithm>

std::string lcs (const std::string &a, const std::string &b) {
    const int m = a.size(), n = b.size();
    int i , j, index, score[m + 1][n + 1];

    for (i = 0; i <= m; i++) {
        for (j = 0; j <= n; j++) {
            if (i == 0 || j == 0)
                score[i][j] = 0;
            else if (a[i - 1] == b[j - 1])
                score[i][j] = score[i - 1][j - 1] + 1;
            else
                score[i][j] = std::max(score[i - 1][j], score[i][j - 1]);
        }
    }

    index = score[m][n];
    i = m, j = n;
    std::string sub(index, 0);

    while (i > 0 && j > 0) {
        if (a[i - 1] == b[j - 1]) {
            sub[--index] = a[i - 1];
            i--, j--;
        } else if (score[i - 1][j] > score[i][j - 1]) {
            i--;
        } else {
            j--;
        }
    }
    return sub;
}

File: 4 Kyu/4 kyu - Longest Common Subsequence (Performance version).rs
fn lcs(x: &str, y: &str) -> String {

    let a = x.chars().collect::<Vec<_>>();
    let b = y.chars().collect::<Vec<_>>();

    let mut m = a.len();
    let mut n = b.len();
    let mut os = String::new();
    let mut score = vec![vec![0;n+1];m+1];

    for i in 0..=m {
        for j in 0..=n {
            if i == 0 || j == 0 {
                score[i][j] = 0;
            } else if a[i - 1] == b[j - 1] {
                score[i][j] = score[i - 1][j - 1] + 1;
            } else {
                score[i][j] = std::cmp::max(score[i - 1][j], score[i][j - 1]);
            }
        }
    }

    while m > 0 && n > 0 {
        if a[m - 1] == b[n - 1] {
            os.push(a[m - 1]);
            m -= 1; n -= 1;
        } else if score[m - 1][n] > score[m][n - 1] {
            m -= 1;
        } else {
            n -= 1;
        }
    }

    os.chars().rev().collect()
